File: src/versions.zig
const std = @import("std");

const io = std.io;
const print = std.debug.print;
const json = std.json;

const Header = std.http.Client.Request.Headers;
const ResponseStorage = std.http.Client.FetchOptions.ResponseStorage;
const FetchOptions = std.http.Client.FetchOptions;
const Location = std.http.Client.FetchOptions.Location;

pub fn fetchPythonVersions() !void {
    const allocator = std.heap.page_allocator;
    var client = std.http.Client{
        .allocator = allocator,
    };

    const pythonGitTagsURI = try std.Uri.parse("https://api.github.com/repos/python/cpython/git/matching-refs/tags/v");

    const headers = Header{};

    var response_raw = std.ArrayList(u8).init(allocator);
    defer response_raw.deinit();

    const fetchOptions = FetchOptions{
        .method = .GET,
        .headers = headers,
        .location = Location{ .uri = pythonGitTagsURI },
        .response_storage = ResponseStorage{ .dynamic = &response_raw },
    };

    const req = try client.fetch(fetchOptions);

    print("{any}\n", .{req.status});

    const parseOpts = .{ .ignore_unknown_fields = true };
    const data = try json.parseFromSliceLeaky(VersionResponse, allocator, response_raw.items, parseOpts);

    print("{any}", .{data.len});

    for (data) |item| {
        const out: []u8 = try allocator.alloc(u8, item.ref.len - 10);
        _ = std.mem.replace(u8, item.ref, "refs/tags/", "", out);
        print("{s}\n", .{out});
    }
}

const VersionResponse = []struct {
    ref: []const u8,
};

File: src/python/versions.rs
use reqwest::{self, header::USER_AGENT};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
struct GitTag {
    #[serde(rename = "ref")]
    tag: String,
}

pub async fn fetch_available_versions() -> Result<Vec<String>, Box<dyn std::error::Error>> {
    let client = reqwest::Client::new();

    let res = client
        .get("https://api.github.com/repos/python/cpython/git/matching-refs/tags/v")
        .header(USER_AGENT, "reqwest rust")
        .send()
        .await?;
    let mut versions: Vec<String> = [].to_vec();

    if res.status() == 200 {
        let tags: Vec<GitTag> = res.json().await?;
        for tag in tags {
            let version = tag.tag.replace("refs/tags/", "");
            if !version.contains("a") && !version.contains("b") && !version.contains("c") {
                versions.push(version);
            }
        }
    }

    versions.reverse();

    return Ok(versions);
}

pub async fn get_latest_minor_of_version(
    major: String,
    minor: String,
) -> Result<String, Box<dyn std::error::Error>> {
    let versions = fetch_available_versions().await?;

    for version in versions {
        let version_vec = version.split(".").filter(|&i| i != "").collect::<Vec<_>>();

        if version_vec[0] == "v".to_string() + &major && version_vec[1] == minor {
            return Ok(version);
        }
    }

    panic!("Unable to find suitable version");
}
